### ISSUE TITLE
XCode shader debugging not working when using RenderStatePackager

### ISSUE THREAD

**Issue by speedym** [DiligentGraphics/DiligentTools#122]
When using XCode frame capture for debugging shaders, double clicking on the shader shows an XCode error message which says source code / debug symbols are not available.

The following is the config .json file I'm using:
{
    "Metal": {
        "CompileOptionsiOS": "-sdk iphoneos metal -g -gembed-source -gline-tables-only -frecord-sources"
    }
}

I am not seeing the source code embedded in the PSO binary files.

**Comment by TheMostDiligent** [comment]
A possible solution is that `-frecord-sources` command line option may allow embedding debug information into the metal lib. It is not however clear if this option works for the library as the example shows how to use it with AIR files. If this is the case, it is possible to embed debug information directly into the archive when the `STRIP_REFLECTION` flag is not used.

Another solution may be feasible: [It appears](https://developer.apple.com/documentation/metal/libraries/generating_and_loading_a_metal_library_symbol_file) that shader symbol files should be generated separately and then loaded by the XCode shader debugging tool (NOTE: this option is only stated to be supported by XCode 13 or later). It also appears that XCode allows specifying a path to the folder containing shader debug symbol files. In this case a solution is for the render state packager to output shader symbol files to a specified folder. The path to the folder will be provided to the packaging tool as an optional parameter (`SerializationDeviceMtlInfo::DebugSymbolPath`) and corresponding option will be added to the render state packager JSON configuration file.

**Comment by speedym** [comment]
I think the `-frecord-sources` would work best for us. How can we manipulate the `STRIP_REFLECTION` flag?

**Comment by TheMostDiligent** [comment]
`STRIP_REFLECTION` flag affects Vulkan/SPIRV. For Metal you can directly provide all flags through the command line and `STRIP_REFLECTION` does not have effect.

`STRIP_REFLECTION` can be set in the archiver using the `strip_reflection` option.

**Comment by TheMostDiligent** [comment]
So it turned out that XCode tells exactly which command line options should be used to enable shader code debugging:

<img width="262" alt="Screen Shot 2022-02-22 at 11 32 17 AM" src="https://user-images.githubusercontent.com/61806567/155205592-4ae7d9cd-8399-4955-8c65-d795d4ba0044.png">

Adding `-gline-tables-only -MO` to the command line options make the XCode show the shader source code.

You can use the following config file:

```json
{
    "Metal": {
        "CompileOptionsMacOS": "-sdk macosx metal -gline-tables-only -MO",
        "CompileOptionsiOS": "-sdk iphoneos metal -gline-tables-only -MO"
    }
}
```

@speedym Can you try these options and see if they work for you?

**Comment by speedym** [comment]
Hmm, interesting. This dialog changed from the last time I saw it (it specifies different options now). Which XCode version are you using?

I have tried the options from the previous version of the dialog, and they didn't work. Have you perhaps tried those new ones out with Diligent examples?

**Comment by TheMostDiligent** [comment]
Yes, I tried and these options work with our examples. Without these options, XCode shows the dialog above.

**Comment by TheMostDiligent** [comment]
I tested with XCode 12.4 (12D4e). I would be shocked if they changed the command line options in newer version.

**Comment by speedym** [comment]
Just for reference this is the dialog I was seeing (I tried adding options from this dialog, but they didn't work):
![shader source not found](https://user-images.githubusercontent.com/2362190/155289063-3f0448da-f52f-4195-aa19-c2a4519c6dce.png)

**Comment by TheMostDiligent** [comment]
What about the options that my XCode suggests?

**Comment by speedym** [comment]
On XCode 13.2.1 our app is getting stuck on the following screen:
<img width="1074" alt="Screenshot 2022-02-23 at 14 25 42" src="https://user-images.githubusercontent.com/2362190/155328463-f8fa7eae-14ad-4743-8066-3bb0efa1f1ef.png">

I'll try it out with XCode 12.4.

**Comment by speedym** [comment]
I managed to resolve the above XCode getting stuck error by upgrading iOS to latest (15.3.1).

**Comment by speedym** [comment]
In the good news, It turned out shader profiling is now working on our side as well! :)

**Comment by speedym** [comment]
The config file which works for latest XCode (13.2.1) is the following:
```
{
    "Metal": {
        "CompileOptionsMacOS": "-sdk macosx metal -gline-tables-only -frecord-sources",
        "CompileOptionsiOS": "-sdk iphoneos metal -gline-tables-only -frecord-sources"
    }
}
```

**Comment by speedym** [comment]
@TheMostDiligent Another note on this one: When compiling shaders with iOS deployment target == 11, shader profiling fails to work even though flags are set up correctly. The workaround for this is to increase the deployment target by using CMAKE_OSX_DEPLOYMENT_TARGET & XCODE_ATTRIBUTE_IPHONEOS_DEPLOYMENT_TARGET, and what worked for me is 13.6. I haven't tried other deployment targets. Maybe it's worth printing a warning during Diligent CMake init about this Apple bug. Also, note that I haven't tested this on MacOS, only on iOS.

**Comment by speedym** [comment]
Just a small addition, earliest deployment target which works for shader profiling is 14.0. Previous deployment (<= 13.6) targets only show the shader sources.
